File: internship_backend_nov24/jury_solutions/python/e.py
def isInBounds(x, y, n, m):
    return x >= 0 and x < n and y >= 0 and y < m

def dfs(startX, startY, setNumber, n, m, setsNumbers, field, used, steps):
    stack = []
    stack.append((startX, startY))
    used[startX][startY] = setNumber

    while(len(stack) != 0):
        (currentX, currentY) = stack[-1]
        stack.pop()
        setsNumbers[setNumber]+=1
        for (stepX, stepY) in steps:
            posX = currentX+stepX
            posY = currentY+stepY
            if isInBounds(posX, posY, n, m) and field[posX][posY] == 'X' and used[posX][posY] == 0:
                stack.append((posX, posY))
                used[posX][posY] = setNumber
                
    

steps = [ (1, 0), (-1, 0), (0, 1), (0, -1) ]

n,m,q = map(int, input().split())
setsNumbers = [0 for i in range(n * m)]
field = [ [0 for i in range(m)] for i in range(n)]
used = [ [0 for i in range(m)] for i in range(n)]

for i in range(n):
    line = input()
    for j in range(m):
        field[i][j] = line[j]

setCounter = 0

for i in range(n):
    for j in range(m):
        if used[i][j] == 0 and field[i][j] == 'X':
            setCounter+=1
            dfs(i, j, setCounter, n, m, setsNumbers, field, used, steps)

for i in range(q):
    x, y = map(int, input().split())
    x-=1
    y-=1
    if used[x][y] == 0:
        print("MISS")
    else:
        setsNumbers[used[x][y]]-=1
        if setsNumbers[used[x][y]] == 0:
            print("DESTROY")
        else:
            print("HIT")

File: internship_backend_nov24/jury_solutions/cs/e.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;


class Program
{
    static void Main()
    {
        Solve();
    }

    static List<(int X, int Y)> steps = new List<(int X, int Y)> { (1, 0), (-1, 0), (0, 1), (0, -1) };
    
    static bool IsInBounds(int x, int y, int n, int m) => x >= 0 && x < n && y >= 0 && y < m;
    
    static void DfsNoRec(int startX, int startY, int setNumber, int n, int m,
        List<int> setsNumbers, 
        char[,] field,
        int[,] used
    )
    {
        Stack<(int X, int Y)> stack = new Stack<(int X, int Y)>();
        stack.Push((startX, startY));
        used[startX, startY] = setNumber;

        while (stack.Count > 0)
        {
            var (currentX, currentY) = stack.Pop();
            setsNumbers[setNumber]++;
            foreach (var position in steps.Select(s => (s.X + currentX, s.Y + currentY)).Where(s =>
                         IsInBounds(s.Item1, s.Item2, n, m) 
                         && field[s.Item1, s.Item2] == 'X' 
                         && used[s.Item1, s.Item2] == 0)
                    )
            {
                used[position.Item1, position.Item2] = setNumber;
                stack.Push((position.Item1, position.Item2));
            }
        }
    }
    
    public static void Solve()
    {
        var nmq = Console.ReadLine().Split();

        var n = int.Parse(nmq[0]);
        var m = int.Parse(nmq[1]);
        var q = int.Parse(nmq[2]);

        var setsNumbers = new List<int>(n * m) { 0 };

        var field = new char[n, m];
        var used = new int[n, m];

        for (var i = 0; i < n; i++)
        {
            var line = Console.ReadLine();
            for (var j = 0; j < m; j++)
                field[i, j] = line[j];
        }

        var setCounter = 0;
        for (var i = 0; i < n; i++)
        for (var j = 0; j < m; j++)
        {
            if (used[i, j] == 0 && field[i, j] == 'X')
            {
                setCounter++;
                setsNumbers.Add(0);
                DfsNoRec(i, j, setCounter, n, m, setsNumbers, field, used);
            }
        }

        while (q-- > 0)
        {
            var position = Console.ReadLine().Split().Select(int.Parse).ToArray();
            var x = position[0] - 1;
            var y = position[1] - 1;
            if (used[x, y] == 0)
                Console.WriteLine("MISS");
            else
            {
                setsNumbers[used[x, y]]--;
                Console.WriteLine(setsNumbers[used[x, y]] == 0 ? "DESTROY" : "HIT");
            }
        }
    }
}